File: packages/mobile_android/app/src/main/java/dev/verdeai/app/Markdown.kt
package dev.verdeai.app

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.*
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import dev.verdeai.core.*

/**
 * Block model built from the core's K-11 markdown AST. There is deliberately no Kotlin markdown
 * parser: when the core cannot render a body (too large, query failure) the source is shown as text.
 */
internal sealed interface MdBlock {
    data class Paragraph(val text: AnnotatedString) : MdBlock
    data class Heading(val level: Int, val text: AnnotatedString) : MdBlock
    data class Bullets(val ordered: Boolean, val items: List<List<MdBlock>>) : MdBlock
    data class Quote(val blocks: List<MdBlock>) : MdBlock
    data class Code(val code: String, val language: String?) : MdBlock
    data object Rule : MdBlock
    data class Table(val rows: List<List<AnnotatedString>>) : MdBlock
}

internal data class MdStyle(val link: Color, val codeBackground: Color)

/** K-11 `highlight` queries for code blocks; the transcript and the file viewer (D-12) supply one. */
internal interface HighlightSource {
    fun cachedHighlight(code: String, language: String): RenderResult<List<RenderSpan>>?
    suspend fun highlight(code: String, language: String): RenderResult<List<RenderSpan>>
}

/** Only the link schemes the core admits are ever made tappable. */
internal fun safeLinkUrl(url: String?): String? {
    val value = url?.trim() ?: return null
    val scheme = value.substringBefore(':', "").lowercase()
    return value.takeIf { scheme == "http" || scheme == "https" || scheme == "mailto" }
}

internal fun citationLabel(citation: FileCitation) = basename(citation.path) +
    (citation.line?.let { line -> ":$line" + (citation.end_line?.takeIf { it > line }?.let { "-$it" } ?: "") } ?: "")

internal fun markdownBlocks(nodes: List<MarkdownNode>, style: MdStyle, onCitation: (FileCitation) -> Unit): List<MdBlock> {
    val out = ArrayList<MdBlock>()
    for (node in nodes) block(node, style, onCitation, out)
    return out
}

private fun block(node: MarkdownNode, style: MdStyle, onCitation: (FileCitation) -> Unit, out: MutableList<MdBlock>) {
    when (node.kind) {
        "document" -> node.children.forEach { block(it, style, onCitation, out) }
        "paragraph" -> out.add(MdBlock.Paragraph(inline(node.children, style, onCitation)))
        "heading" -> out.add(MdBlock.Heading((node.level ?: 1).coerceIn(1, 6), inline(node.children, style, onCitation)))
        "list" -> out.add(MdBlock.Bullets(node.ordered == true, node.children.map { item ->
            val blocks = ArrayList<MdBlock>()
            if (item.kind == "item") item.children.forEach { block(it, style, onCitation, blocks) } else block(item, style, onCitation, blocks)
            blocks
        }))
        "quote" -> out.add(MdBlock.Quote(markdownBlocks(node.children, style, onCitation)))
        "code_block" -> out.add(MdBlock.Code(node.text ?: "", node.language?.takeIf { it.isNotBlank() }))
        "thematic_break" -> out.add(MdBlock.Rule)
        "table" -> out.add(MdBlock.Table(node.children.filter { it.kind == "table_row" }.map { row ->
            row.children.map { cell -> inline(cell.children, style, onCitation) }
        }))
        // Inline content at block level (e.g. a bare image) becomes its own paragraph.
        else -> out.add(MdBlock.Paragraph(inline(listOf(node), style, onCitation)))
    }
}

private fun inline(nodes: List<MarkdownNode>, style: MdStyle, onCitation: (FileCitation) -> Unit): AnnotatedString =
    buildAnnotatedString { nodes.forEach { appendInline(it, style, onCitation) } }

private fun AnnotatedString.Builder.appendInline(node: MarkdownNode, style: MdStyle, onCitation: (FileCitation) -> Unit) {
    fun children() = node.children.forEach { appendInline(it, style, onCitation) }
    val linkStyles = TextLinkStyles(SpanStyle(color = style.link, textDecoration = TextDecoration.Underline))
    when (node.kind) {
        "text" -> append(node.text ?: "")
        "emphasis" -> withStyle(SpanStyle(fontStyle = FontStyle.Italic)) { children() }
        "strong" -> withStyle(SpanStyle(fontWeight = FontWeight.Bold)) { children() }
        "strike" -> withStyle(SpanStyle(textDecoration = TextDecoration.LineThrough)) { children() }
        "code" -> withStyle(SpanStyle(fontFamily = VerdeMono, color = VerdeColors.Heading1, background = style.codeBackground)) { append(node.text ?: "") }
        "line_break" -> append('\n')
        "link" -> {
            val citation = node.citation
            val url = safeLinkUrl(node.url)
            when {
                citation != null -> withLink(LinkAnnotation.Clickable("citation", linkStyles) { onCitation(citation) }) {
                    if (node.children.isEmpty()) append(citationLabel(citation)) else children()
                }
                url != null -> withLink(LinkAnnotation.Url(url, linkStyles)) { children() }
                else -> withStyle(SpanStyle(textDecoration = TextDecoration.Underline)) { children() }
            }
        }
        // Remote images are never fetched; the alt text stands in.
        "image" -> withStyle(SpanStyle(fontStyle = FontStyle.Italic)) {
            append("[image")
            if (node.children.isNotEmpty()) { append(": "); children() }
            append("]")
        }
        else -> children()
    }
}

/**
 * Maps every UTF-8 byte offset of [text] to its UTF-16 index (continuation bytes map to the start
 * of their character), so K-11 byte spans can style a Kotlin string.
 */
internal fun utf8ToUtf16(text: String): IntArray {
    var bytes = 0
    var i = 0
    while (i < text.length) { val cp = text.codePointAt(i); bytes += utf8Length(cp); i += Character.charCount(cp) }
    val out = IntArray(bytes + 1)
    var b = 0
    i = 0
    while (i < text.length) {
        val cp = text.codePointAt(i)
        repeat(utf8Length(cp)) { out[b++] = i }
        i += Character.charCount(cp)
    }
    out[b] = text.length
    return out
}

private fun utf8Length(cp: Int) = when { cp < 0x80 -> 1; cp < 0x800 -> 2; cp < 0x10000 -> 3; else -> 4 }

internal fun highlighted(code: String, spans: List<RenderSpan>, color: (String) -> SpanStyle?): AnnotatedString {
    if (spans.isEmpty()) return AnnotatedString(code)
    val map = utf8ToUtf16(code)
    val last = map.size - 1
    return buildAnnotatedString {
        append(code)
        for (span in spans) {
            val style = color(span.kind) ?: continue
            val start = map[span.start.coerceAtMost(last.toULong()).toInt()]
            val end = map[span.end.coerceAtMost(last.toULong()).toInt()]
            if (end > start) addStyle(style, start, end)
        }
    }
}

@Composable
internal fun tokenStyle(): (String) -> SpanStyle? {
    val c = MaterialTheme.colorScheme
    return remember(c) {
        { kind ->
            when (kind) {
                "keyword" -> SpanStyle(color = Color(0xFFFBC12D), fontWeight = FontWeight.SemiBold)
                "string" -> SpanStyle(color = Color(0xFF4EE29E))
                "number", "constant_name" -> SpanStyle(color = VerdeColors.Heading1)
                "comment" -> SpanStyle(color = c.outline, fontStyle = FontStyle.Italic)
                "type_name" -> SpanStyle(color = c.tertiary)
                "function_name" -> SpanStyle(color = Color(0xFF5ECB83))
                "property_name" -> SpanStyle(color = c.secondary)
                "operator", "punctuation" -> SpanStyle(color = c.onSurfaceVariant)
                else -> null
            }
        }
    }
}

internal const val MARKDOWN_TAG = "markdown"
internal const val PLAIN_TAG = "plain-text"

/** Renders [text] from the core's markdown AST, falling back to the literal source. */
@Composable
internal fun MarkdownText(text: String, model: TranscriptModel, onCitation: (FileCitation) -> Unit, modifier: Modifier = Modifier) {
    val result by produceState(model.cachedMarkdown(text), text) { value = model.markdown(text) }
    val nodes = result?.value
    if (nodes == null) {
        // Pending (first frame) or unrenderable: the source text, never re-parsed in Kotlin.
        Text(text, modifier.testTag(PLAIN_TAG), style = MaterialTheme.typography.bodyMedium)
        return
    }
    val colors = MaterialTheme.colorScheme
    val callback by rememberUpdatedState(onCitation)
    val blocks = remember(nodes, colors) { markdownBlocks(nodes, MdStyle(colors.primary, colors.surfaceVariant)) { callback(it) } }
    MarkdownBlocks(blocks, model, modifier.testTag(MARKDOWN_TAG))
}

@Composable
internal fun MarkdownBlocks(blocks: List<MdBlock>, model: HighlightSource, modifier: Modifier = Modifier) {
    Column(modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) {
        blocks.forEach { MdBlockView(it, model) }
    }
}

@Composable
private fun MdBlockView(block: MdBlock, model: HighlightSource) {
    val type = MaterialTheme.typography
    when (block) {
        is MdBlock.Paragraph -> Text(block.text, style = type.bodyMedium)
        is MdBlock.Heading -> Text(block.text,
            color = when (block.level) { 1 -> VerdeColors.Heading1; 2 -> VerdeColors.Heading2; 3 -> VerdeColors.Heading3; else -> VerdeColors.Heading4 },
            style = (when (block.level) { 1 -> type.headlineMedium; 2 -> type.headlineSmall; else -> type.titleLarge }))
        is MdBlock.Bullets -> Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            block.items.forEachIndexed { index, item ->
                Row {
                    Text(if (block.ordered) "${index + 1}." else "•", Modifier.widthIn(min = 20.dp), style = type.bodyMedium)
                    MarkdownBlocks(item, model, Modifier.weight(1f))
                }
            }
        }
        is MdBlock.Quote -> Row(Modifier.height(IntrinsicSize.Min)) {
            Box(Modifier.width(3.dp).fillMaxHeight().background(MaterialTheme.colorScheme.outlineVariant))
            MarkdownBlocks(block.blocks, model, Modifier.padding(start = 8.dp).weight(1f))
        }
        is MdBlock.Code -> CodeBlock(block.code, block.language, model)
        MdBlock.Rule -> HorizontalDivider()
        is MdBlock.Table -> Column(Modifier.horizontalScroll(rememberScrollState())
            .border(1.dp, MaterialTheme.colorScheme.outlineVariant, RoundedCornerShape(4.dp))) {
            block.rows.forEachIndexed { index, row ->
                Row {
                    row.forEach { cell ->
                        Text(cell, Modifier.width(140.dp).padding(6.dp),
                            style = if (index == 0) type.labelLarge else type.bodySmall)
                    }
                }
                if (index < block.rows.size - 1) HorizontalDivider(Modifier.width(140.dp * row.size))
            }
        }
    }
}

internal const val CODE_BLOCK_TAG = "code-block"

@Composable
internal fun CodeBlock(code: String, language: String?, model: HighlightSource) {
    @Suppress("DEPRECATION") val clipboard = LocalClipboardManager.current
    val spans by produceState(language?.let { model.cachedHighlight(code, it) }, code, language) {
        value = language?.let { model.highlight(code, it) }
    }
    val style = tokenStyle()
    // Spans index the exact code the core highlighted; only the trailing newline is dropped for display.
    val text = remember(code, spans, style) { highlighted(code, spans?.value.orEmpty(), style).let { if (code.endsWith("\n")) it.subSequence(0, code.length - 1) else it } }
    Column(Modifier.fillMaxWidth().background(VerdeColors.Background, RoundedCornerShape(7.dp))
        .border(1.dp, VerdeColors.Border, RoundedCornerShape(7.dp)).testTag(CODE_BLOCK_TAG)) {
        Row(Modifier.fillMaxWidth().background(VerdeColors.PanelAlt, RoundedCornerShape(topStart = 7.dp, topEnd = 7.dp)).padding(start = 10.dp), verticalAlignment = Alignment.CenterVertically) {
            Text(language ?: "code", Modifier.weight(1f), style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            TextButton(onClick = { clipboard.setText(AnnotatedString(code.removeSuffix("\n"))) }) { Text("Copy code") }
        }
        Text(text, Modifier.horizontalScroll(rememberScrollState()).padding(start = 10.dp, end = 10.dp, bottom = 10.dp),
            style = MaterialTheme.typography.bodySmall.copy(fontFamily = VerdeMono), softWrap = false)
    }
}

File: packages/mobile_ios/App/Markdown.swift
import SwiftUI
import UIKit

/// Block model built from the core's K-11 markdown AST. There is deliberately no Swift markdown
/// parser: when the core cannot render a body (too large, query failure) the source is shown as text.
indirect enum MdBlock: Equatable {
    case paragraph(AttributedString)
    case heading(Int, AttributedString)
    case bullets(ordered: Bool, items: [[MdBlock]])
    case quote([MdBlock])
    case code(String, language: String?)
    case rule
    case table([[AttributedString]])

    var kindName: String {
        switch self {
        case .paragraph: return "Paragraph"
        case .heading: return "Heading"
        case .bullets: return "Bullets"
        case .quote: return "Quote"
        case .code: return "Code"
        case .rule: return "Rule"
        case .table: return "Table"
        }
    }
}

/// A finished render query; `value == nil` means "render the source text as-is".
struct RenderResult<T> {
    let value: T?
    init(_ value: T?) { self.value = value }
}

/// K-11 `highlight` queries for code blocks; the transcript (and later the I-09 file viewer) supply one.
@MainActor
protocol HighlightSource: AnyObject {
    func cachedHighlight(_ code: String, language: String) -> RenderResult<[RenderSpan]>?
    func highlight(_ code: String, language: String) async -> RenderResult<[RenderSpan]>
}

/// Only the link schemes the core admits are ever made tappable.
func safeLinkUrl(_ url: String?) -> URL? {
    guard let value = url?.trimmingCharacters(in: .whitespacesAndNewlines), let colon = value.firstIndex(of: ":") else { return nil }
    guard ["http", "https", "mailto"].contains(value[..<colon].lowercased()) else { return nil }
    return URL(string: value)
}

func citationLabel(_ citation: FileCitation) -> String {
    var label = basename(citation.path)
    if let line = citation.line {
        label += ":\(line)"
        if let end = citation.end_line, end > line { label += "-\(end)" }
    }
    return label
}

/// File citations travel inside the attributed text as private `verde-citation:` links and are
/// routed back through the transcript's `OpenURLAction`; they are never opened by the system.
let citationScheme = "verde-citation"

func citationURL(_ citation: FileCitation) -> URL? {
    var components = URLComponents()
    components.scheme = citationScheme
    var items = [URLQueryItem(name: "path", value: citation.path)]
    if let line = citation.line { items.append(URLQueryItem(name: "line", value: String(line))) }
    if let end = citation.end_line { items.append(URLQueryItem(name: "end", value: String(end))) }
    components.queryItems = items
    return components.url
}

func citation(from url: URL) -> FileCitation? {
    guard url.scheme == citationScheme,
          let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems,
          let path = items.first(where: { $0.name == "path" })?.value else { return nil }
    func number(_ name: String) -> UInt64? { items.first { $0.name == name }?.value.flatMap(UInt64.init) }
    return FileCitation(path: path, line: number("line"), end_line: number("end"))
}

// MARK: - AST → blocks

func markdownBlocks(_ nodes: [MarkdownNode]) -> [MdBlock] {
    var out: [MdBlock] = []
    for node in nodes { block(node, into: &out) }
    return out
}

private func block(_ node: MarkdownNode, into out: inout [MdBlock]) {
    switch node.kind {
    case "document": node.children.forEach { block($0, into: &out) }
    case "paragraph": out.append(.paragraph(inline(node.children)))
    case "heading": out.append(.heading(min(max(Int(node.level ?? 1), 1), 6), inline(node.children)))
    case "list":
        out.append(.bullets(ordered: node.ordered == true, items: node.children.map { item in
            var blocks: [MdBlock] = []
            if item.kind == "item" { item.children.forEach { block($0, into: &blocks) } } else { block(item, into: &blocks) }
            return blocks
        }))
    case "quote": out.append(.quote(markdownBlocks(node.children)))
    case "code_block":
        let language = node.language?.trimmingCharacters(in: .whitespaces)
        out.append(.code(node.text ?? "", language: language?.isEmpty == false ? node.language : nil))
    case "thematic_break": out.append(.rule)
    case "table":
        out.append(.table(node.children.filter { $0.kind == "table_row" }.map { row in row.children.map { inline($0.children) } }))
    default:
        // Inline content at block level (e.g. a bare image) becomes its own paragraph.
        out.append(.paragraph(inline([node])))
    }
}

private struct InlineStyle {
    var intent: InlinePresentationIntent = []
    var strike = false
    var underline = false
    var code = false
    var link: URL?
}

private typealias SwiftUIKeys = AttributeScopes.SwiftUIAttributes
private typealias FoundationKeys = AttributeScopes.FoundationAttributes

let inlineCodeBackground = Color(uiColor: .tertiarySystemFill)

private func inline(_ nodes: [MarkdownNode]) -> AttributedString {
    var out = AttributedString()
    for node in nodes { appendInline(node, InlineStyle(), into: &out) }
    return out
}

private func appendInline(_ node: MarkdownNode, _ style: InlineStyle, into out: inout AttributedString) {
    func run(_ text: String, _ s: InlineStyle) {
        guard !text.isEmpty else { return }
        var c = AttributeContainer()
        if !s.intent.isEmpty { c[FoundationKeys.InlinePresentationIntentAttribute.self] = s.intent }
        if s.strike { c[SwiftUIKeys.StrikethroughStyleAttribute.self] = .single }
        if s.underline || s.link != nil { c[SwiftUIKeys.UnderlineStyleAttribute.self] = .single }
        if s.code { c[SwiftUIKeys.BackgroundColorAttribute.self] = inlineCodeBackground }
        if let link = s.link { c[FoundationKeys.LinkAttribute.self] = link }
        out.append(AttributedString(text, attributes: c))
    }
    func children(_ s: InlineStyle) { node.children.forEach { appendInline($0, s, into: &out) } }
    var s = style
    switch node.kind {
    case "text": run(node.text ?? "", style)
    case "emphasis": s.intent.insert(.emphasized); children(s)
    case "strong": s.intent.insert(.stronglyEmphasized); children(s)
    case "strike": s.strike = true; children(s)
    case "code": s.intent.insert(.code); s.code = true; run(node.text ?? "", s)
    case "line_break": run("\n", style)
    case "link":
        if let citation = node.citation, let url = citationURL(citation) {
            s.link = url
            if node.children.isEmpty { run(citationLabel(citation), s) } else { children(s) }
        } else if let url = safeLinkUrl(node.url) {
            s.link = url; children(s)
        } else {
            s.underline = true; children(s)
        }
    case "image":
        // Remote images are never fetched; the alt text stands in.
        s.intent.insert(.emphasized)
        run("[image", s)
        if !node.children.isEmpty { run(": ", s); children(s) }
        run("]", s)
    default: children(style)
    }
}

// MARK: - K-11 spans

/// Maps every UTF-8 byte offset of `text` to its UTF-16 index (continuation bytes map to the start
/// of their character), so K-11 byte spans can style a Swift string. Same table as Android's.
func utf8ToUtf16(_ text: String) -> [Int] {
    var out: [Int] = []
    out.reserveCapacity(text.utf8.count + 1)
    var index = 0
    for scalar in text.unicodeScalars {
        for _ in 0..<UTF8.width(scalar) { out.append(index) }
        index += UTF16.width(scalar)
    }
    out.append(index)
    return out
}

/// A styled range, in UTF-16 indices.
struct DiffRange: Equatable {
    let start: Int
    let end: Int
    let kind: String
}

/// K-11 byte spans mapped to UTF-16 ranges of `text` (clamped; empty ranges dropped).
func highlightRanges(_ text: String, _ spans: [RenderSpan]) -> [DiffRange] {
    if spans.isEmpty { return [] }
    let map = utf8ToUtf16(text)
    let last = UInt64(map.count - 1)
    return spans.compactMap { span in
        let start = map[Int(min(span.start, last))], end = map[Int(min(span.end, last))]
        return end > start ? DiffRange(start: start, end: end, kind: span.kind) : nil
    }
}

struct StyledRange {
    let start: Int
    let end: Int
    let attributes: AttributeContainer
}

/// Builds `text` with attribute ranges (UTF-16 indices on scalar boundaries). Overlaps merge in
/// range order. One sweep, so thousands of highlight spans stay linear.
func styledText(_ text: String, _ ranges: [StyledRange]) -> AttributedString {
    let units = Array(text.utf16)
    let count = units.count
    let valid = ranges.enumerated().compactMap { index, r -> (Int, Int, Int)? in
        let s = min(max(r.start, 0), count), e = min(max(r.end, 0), count)
        return e > s ? (index, s, e) : nil
    }
    if valid.isEmpty { return AttributedString(text) }
    var cuts = Set([0, count])
    for (_, s, e) in valid { cuts.insert(s); cuts.insert(e) }
    let sortedCuts = cuts.sorted()
    let byStart = valid.sorted { $0.1 < $1.1 }
    var next = 0
    var active: [(Int, Int, Int)] = []
    var out = AttributedString()
    for (a, b) in zip(sortedCuts, sortedCuts.dropFirst()) {
        active.removeAll { $0.2 <= a }
        while next < byStart.count && byStart[next].1 <= a { active.append(byStart[next]); next += 1 }
        let segment = String(decoding: units[a..<b], as: UTF16.self)
        if active.isEmpty { out.append(AttributedString(segment)); continue }
        var attributes = AttributeContainer()
        for (index, _, _) in active.sorted(by: { $0.0 < $1.0 }) { attributes = attributes.merging(ranges[index].attributes) }
        out.append(AttributedString(segment, attributes: attributes))
    }
    return out
}

/// Syntax colours for K-11 token kinds (nil: unstyled).
func tokenAttributes(_ kind: String) -> AttributeContainer? {
    var c = AttributeContainer()
    func color(_ value: Color) { c[SwiftUIKeys.ForegroundColorAttribute.self] = value }
    switch kind {
    case "keyword": color(.accentColor); c[FoundationKeys.InlinePresentationIntentAttribute.self] = .stronglyEmphasized
    case "string": color(Color(uiColor: .systemGreen))
    case "number", "constant_name": color(Color(uiColor: .systemOrange))
    case "comment": color(.secondary); c[FoundationKeys.InlinePresentationIntentAttribute.self] = .emphasized
    case "type_name": color(Color(uiColor: .systemPurple))
    case "function_name": color(Color(uiColor: .systemBlue))
    case "property_name": color(Color(uiColor: .systemIndigo))
    case "operator", "punctuation": color(.secondary)
    default: return nil
    }
    return c
}

/// `code` coloured by the core's highlight spans.
func highlighted(_ code: String, _ spans: [RenderSpan]) -> AttributedString {
    styledText(code, highlightRanges(code, spans).compactMap { range in
        tokenAttributes(range.kind).map { StyledRange(start: range.start, end: range.end, attributes: $0) }
    })
}

// MARK: - Views

/// Renders `text` from the core's markdown AST, falling back to the literal source.
struct MarkdownText: View {
    let text: String
    let model: TranscriptModel
    @State private var result: RenderResult<[MdBlock]>?

    var body: some View {
        // A newer body keeps showing the previous render until its own arrives (no plain flash
        // while streaming); a cached render is used at once.
        let current = model.cachedMarkdown(text) ?? result
        Group {
            if let blocks = current?.value {
                MarkdownBlocks(blocks: blocks, source: model).accessibilityIdentifier("markdown")
            } else {
                // Pending (first frame) or unrenderable: the source text, never re-parsed in Swift.
                Text(text).font(.body).textSelection(.enabled).accessibilityIdentifier("plain-text")
            }
        }
        .task(id: text) { result = await model.markdown(text) }
    }
}

struct MarkdownBlocks: View {
    let blocks: [MdBlock]
    let source: HighlightSource

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(blocks.enumerated()), id: \.offset) { _, block in MdBlockView(block: block, source: source) }
        }
    }
}

private struct MdBlockView: View {
    let block: MdBlock
    let source: HighlightSource

    var body: some View {
        switch block {
        case .paragraph(let text):
            Text(text).font(.body).frame(maxWidth: .infinity, alignment: .leading)
        case .heading(let level, let text):
            Text(text).font(level == 1 ? .title2.bold() : level == 2 ? .title3.bold() : .headline)
                .frame(maxWidth: .infinity, alignment: .leading)
                .accessibilityAddTraits(.isHeader)
        case .bullets(let ordered, let items):
            VStack(alignment: .leading, spacing: 4) {
                ForEach(Array(items.enumerated()), id: \.offset) { index, item in
                    HStack(alignment: .firstTextBaseline, spacing: 4) {
                        Text(ordered ? "\(index + 1)." : "•").font(.body).frame(minWidth: 20, alignment: .leading)
                        MarkdownBlocks(blocks: item, source: source)
                    }
                }
            }
        case .quote(let blocks):
            HStack(alignment: .top, spacing: 8) {
                RoundedRectangle(cornerRadius: 1.5).fill(Color(uiColor: .separator)).frame(width: 3)
                MarkdownBlocks(blocks: blocks, source: source)
            }
            .fixedSize(horizontal: false, vertical: true)
        case .code(let code, let language):
            CodeBlock(code: code, language: language, source: source)
        case .rule:
            Divider()
        case .table(let rows):
            ScrollView(.horizontal) {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(Array(rows.enumerated()), id: \.offset) { index, row in
                        HStack(alignment: .top, spacing: 0) {
                            ForEach(Array(row.enumerated()), id: \.offset) { _, cell in
                                Text(cell).font(index == 0 ? .subheadline.weight(.semibold) : .footnote)
                                    .frame(width: 140, alignment: .leading).padding(6)
                            }
                        }
                        if index < rows.count - 1 { Divider() }
                    }
                }
                .overlay(RoundedRectangle(cornerRadius: 4).stroke(Color(uiColor: .separator)))
            }
        }
    }
}

struct CodeBlock: View {
    let code: String
    let language: String?
    let source: HighlightSource
    @State private var spans: RenderResult<[RenderSpan]>?

    private var key: String { (language ?? "") + "\u{0}" + code }

    var body: some View {
        let current = language.flatMap { source.cachedHighlight(code, language: $0) } ?? spans
        // Spans index the exact code the core highlighted; only the trailing newline is dropped for display.
        var text = highlighted(code, current?.value ?? [])
        if code.hasSuffix("\n"), let last = text.characters.indices.last { text.removeSubrange(last...) }
        return VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text(language ?? "code").font(.caption2).foregroundStyle(.secondary)
                Spacer()
                Button("Copy code") { UIPasteboard.general.string = code.hasSuffix("\n") ? String(code.dropLast()) : code }
                    .font(.caption)
            }
            .padding(.horizontal, 10).padding(.vertical, 4)
            ScrollView(.horizontal, showsIndicators: false) {
                Text(text).font(.system(.footnote, design: .monospaced)).fixedSize()
                    .padding(.horizontal, 10).padding(.bottom, 10)
                    .textSelection(.enabled)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color(uiColor: .secondarySystemBackground), in: RoundedRectangle(cornerRadius: 8))
        .accessibilityIdentifier("code-block")
        .task(id: key) {
            guard let language else { return }
            spans = await source.highlight(code, language: language)
        }
    }
}
